x/gov/keeper/msg_server.go

```diff
@@ -55,36 +64,62 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *v1beta1.MsgSubmitP
 		sdk.NewEvent(
 			sdk.EventTypeMessage,
 			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
-			sdk.NewAttribute(sdk.AttributeKeySender, msg.GetProposer().String()),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.GetProposer()),
 		),
 	)
 
-	submitEvent := sdk.NewEvent(types.EventTypeSubmitProposal, sdk.NewAttribute(types.AttributeKeyProposalType, msg.GetContent().ProposalType()))
 	if votingStarted {
-		submitEvent = submitEvent.AppendAttributes(
+		submitEvent := sdk.NewEvent(types.EventTypeSubmitProposal,
 			sdk.NewAttribute(types.AttributeKeyVotingPeriodStart, fmt.Sprintf("%d", proposal.ProposalId)),
 		)
+
+		ctx.EventManager().EmitEvent(submitEvent)
 	}
 
-	ctx.EventManager().EmitEvent(submitEvent)
-	return &v1beta1.MsgSubmitProposalResponse{
+	return &v1beta2.MsgSubmitProposalResponse{
 		ProposalId: proposal.ProposalId,
 	}, nil
 }
 
-func (k msgServer) Vote(goCtx context.Context, msg *v1beta1.MsgVote) (*v1beta1.MsgVoteResponse, error) {
+func (k msgServer) ExecLegacyContent(goCtx context.Context, msg *v1beta2.MsgExecLegacyContent) (*v1beta2.MsgExecLegacyContentResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	content, err := v1beta2.LegacyContentFromMessage(msg)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(v1beta1.ErrInvalidProposalContent, "%+v", err)
+	}
+
+	// if so ensure that the content has a respective handler
+	if !k.Keeper.legacyRouter.HasRoute(content.ProposalRoute()) {
+		return nil, sdkerrors.Wrap(types.ErrNoProposalHandlerExists, content.ProposalRoute())
+	}
+
+	// Execute the proposal content in a new context branch (with branched store)
+	// to validate the actual parameter changes before the proposal proceeds
+	// through the governance process. State is not persisted.
+	cacheCtx, _ := ctx.CacheContext()
```

Comment:
Don't we eventually want to run `writeCache`. Currently this looks like it never actually persists state. Is this intentional?

Reply:
Ah you're right.

So the way it currently works is:
- in BeginBlock, we create a CacheContext
- if all Msgs pass then we write

Which means that in the msg server implementation of ExecLegacyContent, we don't need a CacheContext at all. I removed it.